src/dbnode/persist/fs/cross_block_reader.go

```diff
@@ -63,51 +66,66 @@ func NewCrossBlockReader(dataFileSetReaders []DataFileSetReader) (CrossBlockRead
 		previousStart = currentStart
 	}
 
-	return &crossBlockReader{dataFileSetReaders: dataFileSetReaders, activeReadersCount: len(dataFileSetReaders)}, nil
+	return &crossBlockReader{
+		dataFileSetReaders: dataFileSetReaders,
+		records:            make([]BlockRecord, 0, len(dataFileSetReaders)),
+	}, nil
 }
 
-func (r *crossBlockReader) Read() (id ident.ID, tags ident.TagIterator, datas []checked.Bytes, checksums []uint32, err error) {
+func (r *crossBlockReader) Next() bool {
 	if !r.initialized {
-		r.initialized = true
 		err := r.init()
 		if err != nil {
-			return nil, nil, nil, nil, err
+			r.err = err
+			return false
 		}
 	}
 
+	if len(r.minHeap) == 0 {
+		return false
+	}
+
 	firstEntry, err := r.readOne()
 	if err != nil {
-		return nil, nil, nil, nil, err
+		r.err = err
+		return false
 	}
 
-	datas = make([]checked.Bytes, 0, r.activeReadersCount)
-	checksums = make([]uint32, 0, r.activeReadersCount)
+	r.id = firstEntry.id
+	r.tags = firstEntry.tags
 
-	datas = append(datas, firstEntry.data)
-	checksums = append(checksums, firstEntry.checksum)
+	r.records = r.records[:0]
```

Comment:
nit: Before we shrink the slice we should release refs to any pointers in the slice. If you don't, they still become "reachable" in view of the garbage collector and hold strong references to say the checked bytes still referenced by the backing array.

```go
// use empty var in inner loop with "for i := range" to have compiler use memclr optimization
// see: https://codereview.appspot.com/137880043
var emptyRecord BlockRecord 
for i := range r.records {
  r.records[i] = emptyRecord
}
```